### PR TITLE
No longer automatically iterate on lookups

### PR DESCRIPTION
**Pull request by samscott89**
PR checklist:

- [x] Python
- [x] Ruby
- [x] Java
- [x] Node.js
- [x] Rust
- [x] Update docs examples
- [x] Doc entry
- [x] Added changelog entry.
- [x] Update FFI support doc to add NextExternal
- [x] New tests for in behaviour
- [ ] Update (out of repository) examples - do last

### CODE REVIEW

File: languages/python/oso/polar/query.py [hunk 1]
```diff
@@ -88,47 +86,33 @@ def handle_make_external(self, data):
 
     def handle_external_call(self, data):
         call_id = data["call_id"]
-        if call_id not in self.calls:
```
Comment (leina05) [inline languages/python/oso/polar/query.py:91]: Can we get rid of `self.calls` entirely now?

File: languages/python/oso/polar/query.py [hunk 2]
```diff
@@ -88,47 +86,33 @@ def handle_make_external(self, data):
...
-            value = next(self.calls[call_id])
-            self.ffi_query.call_result(call_id, self.host.to_polar(value))
-        except StopIteration:
+            attr = getattr(instance, attribute)
+        except AttributeError as e:
+            self.ffi_query.application_error(str(e))
             self.ffi_query.call_result(call_id, None)
+            return
+        if (
+            callable(attr) and not data["args"] is None
```
Comment (leina05) [inline languages/python/oso/polar/query.py:101]: I know this isn't related to your changes, but... 
So `data["args"] is None` is how we check whether or not this was parsed as a field or method? If it's a method with no args, is it an empty list? Should we add an explicit flag here, rather than relying on the `None` check?

Comment (leina05) [reply languages/python/oso/polar/query.py:101]: It's especially weird for that to be the check for methods, since we now have `kwargs`, which will be `None` if you don't have `kwargs`, even if the call is a method.

Comment (samscott89) [reply languages/python/oso/polar/query.py:101]: > I know this isn't related to your changes, but... 
> So `data["args"] is None` is how we check whether or not this was parsed as a field or method? If it's a method with no args, is it an empty list? Should we add an explicit flag here, rather than relying on the `None` check?

All of this is correct.

- Yes that's how we check
- Yes empty list for call instead of none
- a flag might be better. You would still want/need to check the none/list case but might make it less error prone?

Comment (leina05) [reply languages/python/oso/polar/query.py:101]: Yeah, would also improve readability.

File: polar-core/src/vm.rs [hunk 3]
```diff
@@ -1418,6 +1425,27 @@ impl PolarVirtualMachine {
...
                         )?;
                     }
+                    // Push an `ExternalLookup` goal for external instances and built-ins.
+                    Value::Dictionary(_) | Value::ExternalInstance(_) | Value::String(_) => {
+                        // Generate symbol for next result and bind to `false` (default)
+                        let next = self.kb.read().unwrap().gensym("next_value");
+                        self.bind(&next, Term::new_temporary(Value::Boolean(false)));
+                        let next_term = Term::new_temporary(Value::Variable(next.clone()));
+                        let call_id = self.new_call_id(&next);
+
```
Comment (leina05) [inline polar-core/src/vm.rs:1435]: Kind of confused about exactly what's happening here, but I think this is just setting up a variable to receive the next value from the iterator?

File: polar-core/src/vm.rs [hunk 4]
```diff
@@ -1772,10 +1802,10 @@ impl PolarVirtualMachine {
                 self.bind(&answer, Term::new_temporary(Value::Boolean(false)));
 
                 // append unify goal to be evaluated after external op result is returned & bound
-                self.append_goals(vec![Goal::Unify {
+                self.push_goal(Goal::Unify {
```
Comment (leina05) [inline polar-core/src/vm.rs:1805]: Reason for this change?

File: polar-core/benches/bench.rs [hunk 5]
```diff
@@ -169,91 +167,10 @@ pub fn many_rules(c: &mut Criterion) {
     });
 }
```
Comment (leina05) [inline polar-core/benches/bench.rs:169]: I have no idea how bench tests work so not reviewing these changes.

File: polar-core/src/vm.rs [hunk 6]
```diff
@@ -1418,6 +1425,27 @@ impl PolarVirtualMachine {
...
+                        self.bind(&next, Term::new_temporary(Value::Boolean(false)));
+                        let next_term = Term::new_temporary(Value::Variable(next.clone()));
+                        let call_id = self.new_call_id(&next);
+
+                        // append unify goal to be evaluated after
+                        // next result is fetched
+                        self.append_goals(vec![
+                            Goal::NextExternal {
+                                call_id,
+                                term: self.deep_deref(&list),
```
Comment (leina05) [inline polar-core/src/vm.rs:1441]: I think we should rename `list` to reflect that it could be a non-list. Maybe `iterator` would be better.

File: polar-core/src/events.rs [hunk 7]
```diff
@@ -88,4 +88,9 @@ pub enum QueryEvent {
         operator: Operator,
         args: TermList,
     },
+
+    NextExternal {
+        call_id: u64,
+        term: Term,
```
Comment (leina05) [inline polar-core/src/events.rs:94]: The name `term` is very vague. Would be nice if it provided more context, e.g. `iterable_term` or something.

File: languages/java/oso/src/main/java/com/osohq/oso/Query.java [hunk 8]
```diff
@@ -76,7 +118,23 @@ private void handleCall(
       ffiQuery.callResult(callId, null);
       return;
     }
+  }
+
+  /** Helper for `NextExternal` query events */
+  private void handleNextExternal(long callId, JSONObject term) throws Exceptions.OsoException {
+    if (!calls.containsKey(callId)) {
+      Object result = host.toJava(term);
```
Comment (leina05) [inline languages/java/oso/src/main/java/com/osohq/oso/Query.java:126]: Isn't it possible that `term` is an instance, in which case we would need to use `host.getInstance()`, not `host.toJava()` (the same way we do on lines 70-80?

File: languages/java/oso/src/main/java/com/osohq/oso/Query.java [hunk 9]
```diff
@@ -66,7 +66,49 @@ private void handleCall(
...
     try {
-      registerCall(attrName, args, callId, polarInstance);
+      Object instance;
+      if (polarInstance.getJSONObject("value").has("ExternalInstance")) {
+        long instanceId =
+            polarInstance
+                .getJSONObject("value")
+                .getJSONObject("ExternalInstance")
+                .getLong("instance_id");
+        instance = host.getInstance(instanceId);
```
Comment (leina05) [inline languages/java/oso/src/main/java/com/osohq/oso/Query.java:76]: Can this code be folded into `host.toJava()`?

Comment (leina05) [reply languages/java/oso/src/main/java/com/osohq/oso/Query.java:76]: After looking at `toJava()`, I think that this branch may be unnecessary because `toJava()` checks for `ExternalInstance` objects. Unless the structure of this event is a special case.

File: languages/java/oso/src/main/java/com/osohq/oso/Query.java [hunk 10]
```diff
@@ -76,7 +118,23 @@ private void handleCall(
...
     }
+  }
+
+  /** Helper for `NextExternal` query events */
+  private void handleNextExternal(long callId, JSONObject term) throws Exceptions.OsoException {
+    if (!calls.containsKey(callId)) {
+      Object result = host.toJava(term);
+      Enumeration<Object> enumResult;
+      if (result instanceof Enumeration) {
+        // TODO: test this
```
Comment (leina05) [inline languages/java/oso/src/main/java/com/osohq/oso/Query.java:129]: Highlighting this TODO

File: languages/java/oso/src/main/java/com/osohq/oso/Query.java [hunk 11]
```diff
@@ -66,7 +66,49 @@ private void handleCall(
...
+      if (polarInstance.getJSONObject("value").has("ExternalInstance")) {
+        long instanceId =
+            polarInstance
+                .getJSONObject("value")
+                .getJSONObject("ExternalInstance")
+                .getLong("instance_id");
+        instance = host.getInstance(instanceId);
+      } else {
+        instance = host.toJava(polarInstance);
+      }
```
Comment (leina05) [inline languages/java/oso/src/main/java/com/osohq/oso/Query.java:79]:
```suggestion
      Object instance = host.toJava(polarInstance);
```

I believe this works. Small clean-up.

File: languages/java/oso/src/test/java/com/osohq/oso/PolarTest.java [hunk 12]
```diff
@@ -288,17 +288,6 @@ public void testNoKeywordArgs() throws Exception {
         () -> p.query("x = (new MyClass(\"test\", 1)).foo(\"test\", id: 1)"));
   }
 
-  @Test
```
Comment (leina05) [inline languages/java/oso/src/test/java/com/osohq/oso/PolarTest.java:291]: Do we need a test to replace this? `testHandleCall`?

**Review by leina05 — COMMENTED** [review]
## Review of only docs, examples,  java, and polar-core
Looks good! High-level, do we need to add tests for the new operand types for `in` (Dictionaries, Strings, and Iterable External Instances) the way we currently have specific tests for lists?

I will review the rest later this afternoon.

File: languages/java/oso/src/main/java/com/osohq/oso/Query.java [hunk 13]
```diff
@@ -66,7 +66,49 @@ private void handleCall(
       args = Optional.of(host.polarListToJava(jArgs.get()));
     }
     try {
-      registerCall(attrName, args, callId, polarInstance);
+      Object instance;
```
[the remark this answers is not in the document]
Comment (samscott89) [reply languages/java/oso/src/main/java/com/osohq/oso/Query.java:69]: It's no longer registering anything. A call always returns a single value. We could keep it as a separate method to do the call part though if you think that would be better?

File: languages/java/oso/src/main/java/com/osohq/oso/Query.java (hunk 11 above)
Comment (samscott89) [reply languages/java/oso/src/main/java/com/osohq/oso/Query.java:79]: Big cleanup! Nice! I was just moving the code that was already there.

File: docs/changelogs/vNEXT.rst [hunk 14]
```diff
@@ -13,12 +13,28 @@ Breaking changes
...
 
-- summary of breaking change
+Calling methods and accessing attributes on application instances no longer
+returns multiple values. Instead, to iterate over values of an application
+instance the explicit ``in`` operator must be used.
 
-Link to migration guide
+**Old syntax**:
+
+   amount = recipe.ingredients().quantity
```
[the remark this answers is not in the document]
Comment (samscott89) [reply docs/changelogs/vNEXT.rst:25]: Yep! Here's a test case where it was used previously: https://github.com/osohq/oso/pull/485/files#diff-347a4ce26269e34690f5914f1f082bccb9868128ebd7caf088a77ea9f5018ed0L23-R24

File: polar-core/src/vm.rs [hunk 15]
```diff
@@ -357,8 +360,7 @@ impl PolarVirtualMachine {
                 call_id,
                 instance,
                 field,
-                check_errors,
```
[the remark this answers is not in the document]
Comment (samscott89) [reply polar-core/src/vm.rs:358]: Previously we needed to push a check error goal after every new LookupExternal event we created. Since there is only ever one LookupExternal event, the CheckError goal can just be pushed directly by the goal creator.

[the remark this answers is not in the document]
Comment (samscott89) [reply polar-core/src/vm.rs:358]: Which is what is happening here: https://github.com/osohq/oso/pull/485/files#diff-31c6da6cb2dee04c7bdefb96252140c683d8d697de90ec8ae7a6745bf46139d9R1635

File: polar-core/src/vm.rs (hunk 3 above)
Comment (samscott89) [reply polar-core/src/vm.rs:1440]: Yup! Same pattern we've used elsewhere. We might want to extract this into a helper method.

File: polar-core/src/vm.rs (hunk 4 above)
Comment (samscott89) [reply polar-core/src/vm.rs:1810]: It seemed wasteful to create a vector to push a single goal, when we have a method that accepts a single goal.

File: docs/using/polar-syntax.rst [hunk 16]
```diff
@@ -402,19 +402,25 @@ arguments must come before keyword arguments::
...
 Will bind ``x`` to ``1``, ``2``, ``3``, in turn, and check that ``x = 1``
 for each. This expression will only succeed for the first item (``1``).
+The left-hand side does not need to be a variable, for example::
+
+    1 in [1, 2, 3, 1]
+
+Will succeed *twice*: 1 is in the first and fourth position.
 
-The ``in`` operator generates *alternatives* for each element of the list.
+The ``in`` operator generates *alternatives* for each element.
```
[the remark this answers is not in the document]
Comment (samscott89) [reply docs/using/polar-syntax.rst:426]: Removing

File: polar-core/src/vm.rs (hunk 3 above)
Comment (samscott89) [reply polar-core/src/vm.rs:1440]: did this `new_call_var`

File: polar-core/benches/bench.rs (hunk 5 above)
Comment (samscott89) [reply polar-core/benches/bench.rs:169]: They were a mess and overly complicated. So I moved the one that benched externals to be a bench in the oso library instead, so that its way simpler.

File: polar-core/src/events.rs (hunk 7 above)
Comment (samscott89) [reply polar-core/src/events.rs:94]: Sure

File: languages/python/oso/polar/query.py (hunk 2 above)
Comment (samscott89) [reply languages/python/oso/polar/query.py:102]: If its okay, I would prefer to deal with this in a follow up API cleanup.

File: languages/java/oso/src/main/java/com/osohq/oso/Query.java (hunk 9 above)
Comment (samscott89) [reply languages/java/oso/src/main/java/com/osohq/oso/Query.java]: I think all of these are handled by your suggestion?

File: languages/java/oso/src/main/java/com/osohq/oso/Query.java (hunk 10 above)
Comment (samscott89) [reply languages/java/oso/src/main/java/com/osohq/oso/Query.java]: Done

File: polar-core/src/vm.rs (hunk 6 above)
Comment (samscott89) [reply polar-core/src/vm.rs]: Done

File: languages/python/oso/polar/query.py (hunk 2 above)
Comment (leina05) [reply languages/python/oso/polar/query.py:102]: Cool, made a notion task

File: languages/python/oso/tests/test_polar.py [hunk 17]
```diff
@@ -835,3 +836,23 @@ class Post:
 
     with pytest.raises(StopIteration):
         next(results)
+
+
+def test_iterators(polar, qeval, qvar):
+    assert qvar('x in "abc"', "x") == ["a", "b", "c"]
+    assert qvar("x in {a: 1, b: 2}", "x") == ["a", "b"]
```
Comment (leina05) [inline languages/python/oso/tests/test_polar.py:843]: It's unfortunate that these features are totally language-specific, even though they appear to be builtin.

File: languages/ruby/lib/oso/polar/query.rb [hunk 18]
```diff
@@ -97,13 +70,33 @@ def application_error(message)
...
-      rescue InvalidCallError => e
+      rescue ArgumentError, NoMethodError => e
         application_error(e.message)
         call_result(nil, call_id: call_id)
+      end
+
+      def handle_next_external(call_id, iterable)
+        unless calls.key? call_id
+          value = host.to_ruby iterable
+          raise InvalidIteratorError unless value.is_a? Enumerable
```
Comment (leina05) [inline languages/ruby/lib/oso/polar/query.rb:93]: Is `Enumerable` really the only check here? Not familiar enough with Ruby.

File: languages/rust/oso/tests/test_polar_rust.rs [hunk 19]
```diff
@@ -489,6 +489,7 @@ fn test_unify_externals() {
         if m == "cannot bind expression 'x = new Foo(1)' to '_instance_23'"));
 }
 
+#[ignore = "Re-enable after updating iterator implementation"]
```
Comment (leina05) [inline languages/rust/oso/tests/test_polar_rust.rs:492]: Does this need to be re-enabled?

**Review by leina05 — COMMENTED** [review]
Looks great, thanks for adding tests. Things I'm thinking about (maybe as follow-ups):
- Super annoying comment, but `NextExternal` as an event name is a little hard to parse without context. `NextIterableItem` or something like that would be more clear, but this is pretty nit picky.
- Are there any public examples we need to update that aren't tested in CI? Thinking about Polar Adventure, Zenefits demo (I already have the updates for this locally), anything else? The `=` to `in` change is breaking, so let's make sure examples are up to date.
- The behavior of `in` for various pseudo-builtin types is pretty different depending on the host language. Not sure if this is something we need to document, or whether it's intuitive enough based on the host language behavior.

If we generate a list of any other repos/examples we need to update for this change and figure out who's responsible for making them before Monday, then I think this is good to go.

File: test/test.polar [hunk 20]
```diff
@@ -81,3 +81,5 @@ testUnifyClass(A);
 testUnspecializedRuleOrder(_x, _y, z) if z = 3;
 testUnspecializedRuleOrder(_x, _y: String, z) if z = 2;
 testUnspecializedRuleOrder(_x: String, _y, z) if z = 1;
+
+testIterables() if 1 in new List([3, 2, 1]);
```
[the remark this answers is not in the document]
Comment (samscott89) [reply test/test.polar:85]: Yep! This just seemed like one way to test the builtin version of list in each language

File: languages/ruby/lib/oso/polar/query.rb (hunk 18 above)
Comment (samscott89) [reply languages/ruby/lib/oso/polar/query.rb:93]: As far as I can tell, that's the main trait. @gj ?

File: languages/rust/oso/tests/test_polar_rust.rs (hunk 19 above)
Comment (samscott89) [reply languages/rust/oso/tests/test_polar_rust.rs:492]: After the Rust changes are implemented (follow up PR incoming)

**Comment by samscott89** [comment]
> 
> 
> Looks great, thanks for adding tests. Things I'm thinking about (maybe as follow-ups):
> 
>   * Super annoying comment, but `NextExternal` as an event name is a little hard to parse without context. `NextIterableItem` or something like that would be more clear, but this is pretty nit picky.
> 

Agree, but would rather address as follow up... For now wanted it consistent

>   * Are there any public examples we need to update that aren't tested in CI? The `=` to `in` change is breaking, so let's make sure examples are up to date.
>       
>       * Polar Adventure
>       * oso-social app (definitely the branch we sent to Zenefits will break from this)
>       * anything else?
> 

Yes, want to do a review of all the repos as a follow up to this. The demo repo will definitely break

>     * The behavior of `in` for various pseudo-builtin types is pretty different depending on the host language. Not sure if this is something we need to document, or whether it's intuitive enough based on the host language behavior.

Agreed. This was annoying me. Just pushed a change to make it work in the core for strings + dicts.

> 
> 
> If we generate a list of any other repos/examples we need to update for this change and figure out who's responsible for making them before Monday, then I think this is good to go.

Sounds good.

File: polar-core/benches/bench.rs (hunk 5 above)
Comment (samscott89) [reply polar-core/benches/bench.rs:193]: In case you're interested, this is how it looks now:
https://github.com/osohq/oso/pull/515/files#diff-ccf3c4f853abe6996692d38527d164a48366a51ed18810306292370442de2abdR73

Instead of all that fake external stuff, we can just create a new struct with whatever implementation we want

File: languages/java/oso/src/test/java/com/osohq/oso/PolarTest.java [hunk 21]
```diff
@@ -562,4 +551,50 @@ public void testPathMapper() throws Exception {
...
+    }
+
+    // code for data structure
+    public Iterator<Integer> iterator() {
+      return list.iterator();
+    }
+  }
+
+  @Test
+  public void testIterators() throws Exception {
```
[the remark this answers is not in the document]
Comment (leina05) [reply languages/java/oso/src/test/java/com/osohq/oso/PolarTest.java:582]: @samscott89

[the remark this answers is not in the document]
Comment (samscott89) [reply languages/java/oso/src/test/java/com/osohq/oso/PolarTest.java:582]: Oh sorry forgot to reply. Enumerations are tested through the prior functionality in `testEnumerationCallResults`.

File: languages/ruby/lib/oso/polar/query.rb (hunk 18 above)
Comment (leina05) [reply languages/ruby/lib/oso/polar/query.rb:93]: Cursory google search says that's correct

**Review by leina05 — APPROVED** [review]

File: languages/ruby/lib/oso/polar/query.rb (hunk 18 above)
Comment (gj) [reply languages/ruby/lib/oso/polar/query.rb:93]: I think this is fine -- the call to `.lazy` on 95 will turn any `Enumerable` into a lazy enumerator. So this check is basically asking "is it possible to create an enumerator out of `value`, even if it isn't already an enumerator?"

File: languages/java/oso/src/main/java/com/osohq/oso/Query.java [hunk 22]
```diff
@@ -76,7 +109,27 @@ private void handleCall(
       ffiQuery.callResult(callId, null);
       return;
     }
+  }
+
+  /** Helper for `NextExternal` query events */
+  private void handleNextExternal(long callId, JSONObject iterable) throws Exceptions.OsoException {
+    if (!calls.containsKey(callId)) {
+      Object result = host.toJava(iterable);
+      Enumeration<Object> enumResult;
```
Comment (gj) [inline languages/java/oso/src/main/java/com/osohq/oso/Query.java:118]: @samscott89 I see there was already some usage of `Enumeration<Object>` in the previous code, but did you think about using Iterator instead?

https://docs.oracle.com/javase/10/docs/api/java/util/Iterator.html

> An iterator over a collection. **Iterator takes the place of Enumeration in the Java Collections Framework.** Iterators differ from enumerations in two ways:
>    - Iterators allow the caller to remove elements from the underlying collection during the iteration with well-defined semantics.
>    - Method names have been improved.
